connect/api/src/main/java/org/apache/kafka/connect/source/SourceTask.java

```diff
@@ -88,18 +89,46 @@ public void commit() throws InterruptedException {
 
     /**
      * <p>
-     * Commit an individual {@link SourceRecord} when the callback from the producer client is received, or if a record is filtered by a transformation.
+     * Commit an individual {@link SourceRecord} when the callback from the producer client is received.
+     * </p>
+     * <p>
+     * This is an alias for {@link commitRecord(SourceRecord, RecordMetadata)} for backwards compatibility. The default
+     * implementation of {@link commitRecord(SourceRecord, RecordMetadata)} just calls this method.
      * </p>
      * <p>
      * SourceTasks are not required to implement this functionality; Kafka Connect will record offsets
      * automatically. This hook is provided for systems that also need to store offsets internally
      * in their own system.
      * </p>
      *
-     * @param record {@link SourceRecord} that was successfully sent via the producer.
+     * @param record {@link SourceRecord} that was successfully sent via the producer or filtered by a transformation
      * @throws InterruptedException
+     * @see commitRecord(SourceRecord, RecordMetadata)
      */
     public void commitRecord(SourceRecord record) throws InterruptedException {
         // This space intentionally left blank.
     }
+
+    /**
+     * <p>
+     * Commit an individual {@link SourceRecord} when the callback from the producer client is received.
+     * </p>
+     * <p>
+     * SourceTasks are not required to implement this functionality; Kafka Connect will record offsets
+     * automatically. This hook is provided for systems that also need to store offsets internally
+     * in their own system.
+     * </p>
+     * <p>
+     * The default implementation just calls @{link commitRecord(SourceRecord)}, which is a nop by default.
+     * </p>
+     *
+     * @param record {@link SourceRecord} that was successfully sent via the producer
+     * @param metadata {@link RecordMetadata} record metadata returned from the broker
```

Comment:
@ryannedolan, we need to clearly specify that `metadata` parameter can be null, and then in the JavaDoc above specify what this means for the source task, namely that a transform dropped/skipped the record and it was not written to Kafka. 

IMO this is necessary so that developers of connector implementations know what the behavior is so they can properly implement their task. (The JavaDoc was not in [KIP-416](https://cwiki.apache.org/confluence/display/KAFKA/KIP-416%3A+Notify+SourceTask+of+ACK%27d+offsets%2C+metadata).)

I also think that it's also worth mentioning here that `SourceTask` implementations need only implement this method *or* the older `commitRecord(SourceRecord)` *or* neither method, but that generally they do not need to implement both since Connect will only call this method. Again, this will help developers that are implementing their own `SourceTask` what they need to do.

Reply:
@rhauch I improved the javadocs further. Should be clear now.